text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,472 @@
+
+Overall, FPC can be considered an addition to Fabric wherein all chaincode computation relies only on the correctness of data provided by an authenticated Client or computed inside of and signed by a validated Trusted Execution Environment. It assumes that the Ordering Service is a trusted system element. The Endorsing Peer outside of the Trusted Execution Environment, however, is considered untrusted.
+
+Writing chaincode for FPC should come natural to developers familiar with Fabric as the programming model (e.g., chaincode lifecycle, chaincode invocations and state) is the same as for normal Fabric chaincode.  The main differences are a (for now at least) different programming language (C++) and a Shim API which, on the one hand, is somewhat simplified but, on the other hand, also contains FPC-specific extensions such as allowing for private as well as public ledger state as well as a secure (yet “replicable”) random-number generator.
+See the [FPC Shim](#fpc-shim) section for more information.
+
+In general, a developer can write chaincode largely without having to be aware that the code executes within a Trusted Execution Environment (TEE), often referred to as an Enclave. However, to understand the architecture and security underpinning it we introduce in the following several new concepts and terms. Except where noted otherwise, all elements of the architecture described below reside on a Fabric Peer.
+
+
+## Terminology:
+
+* Attestation: An important cryptographic feature of Trusted Execution Environments by which the hardware can reliably state exactly what software it is running. This statement is signed by the hardware so that anyone reading it can verify that the statement came from an actual valid and up-to-date TEE. The Attestation can cover both the software portion of the TEE itself as well as any program to be run in it. This attestation takes the form of a "quote" containing (among other things) a measurement of the code and data in the TEE and its software version number, which is signed by the TEE and can be verified by anyone using well known Public Key Infrastructure technology.
+
+* Trusted Execution Environment (TEE): The isolated secure environment in which programs run in encrypted memory, unreadable even by privileged users or system processes. FPC chaincodes run in TEEs.
+
+* Enclave: The TEE technology used for the initial release of FPC will be Intel SGX.  In SGX terminology a TEE is called an _enclave_.  In this document and in general, the terms TEE and Enclave are considered interchangeable.  Intel SGX is the first TEE technology supported as, to date, it is the only TEE with mature support for remote attestation as required by the FPC integrity architecture.  However, our architecture is generic enough to also allow other implementations based on AMD SEV-SNP, ARM TrustZone, or other TEEs.
```

Comment:
> it is the only TEE with mature support for remote attestation as required by the FPC integrity architecture.

I would remove this. This is time-dependent and also a matter of opinion. Someone from AMD, or the NSA may have other opinions.

Reply:
Good point about time-dependency. BTW: the reason why right we think SGX is the only one is that you need security also against (some) hardware adversary (ruling out, e.g., TPM) and it has to provide publicly verifiable attestation (ruling out available AMD SEV; SEV-SNP probably should address that but details are not yet all out and is not yet available).  Curious, though, what did you have in mind when you mentioned NSA?

Reply:
[Historically](https://en.wikipedia.org/wiki/Differential_cryptanalysis), national security agencies have not disclosed vulnerabilities for periods of time in order to have competitive advantage.

Reply:
Ah, ok, you meant that  NSA might be of the opinion that nonesuch exists rather than they had another technology in mind (which i thought was where you were heading at) ...